tests/x11/gnomecase/change_password.pm

```diff
@@ -144,7 +145,9 @@ sub run {
     assert_and_click 'displaymanager-test';
     assert_screen "testUser-login-dm";
     type_string "$pwd4newUser\n";
-    assert_screen "generic-desktop", 120;
+    my @desktop_needles = qw(generic-desktop);
```

Comment:
Please, extract this logic into a function. This is what I wanted to avoid, having patches everywhere that can be changed and not kept sync.

Reply:
So, how should a useful function look like here? Most of the logic is already encapsulated by `opensuse_welcome_applicable`. Should the function just return applicable needle, or also run the assert_screen - then we have to pass the timeout parameter as well.

Wrapping every single check in a specialized function makes the code harder to read, as you can no longer see whats going on. Passing parameters to functions used only in very few places makes it hard to guess what the parameters are for.

Reply:
Definition draft
```perl
sub handle_opensuse_welcome_popup {
  my (%args) = @_;
  $args{timeout} //= 120

  if (opensuse_welcome_applicable() && not_root) {
    assert_screen('opensuse-welcome', $args{timeout});
  }
}
```

Call:
```bash
handle_opensuse_welcome_popup(timeout => 120);
```

Reply:
Sorry, but this quite broken:
The function would have to be:
```assert_desktop_or_opensuse_welcome_for_non_root_user```

The secondary users are never root, so this does not apply here. You could make this a (default) parameter, `user_is_root` and then handle that in the function, iff you want to make the function "universal".

You override the default assert_screen timeout, which is bad practice.

From the function name, I can not see what the function is supposed to do - `handle` is obviously wrong, as it does not "handle" anything. Also, 'generic-desktop' is valid here as well.

When you got everything right, you end up with a complicated 10+ lines function which makes the tests actually harder to read, just to save 1 line each in 2 tests.

**Iff** there is a common pattern, repeated in several places, I am all for abstracting it away. Doing it upfront is IMHO not a good idea.

Reply:
It is a draft, you ask me how that function would look like. I haven't written the function that you have to copy&paste.

> Sorry, but this quite broken:
> The function would have to be:
> `assert_desktop_or_opensuse_welcome_for_non_root_user`

Maybe I understood the popup wrong. Is it not shown only and always the first time a user logs in into a graphical session?


> The secondary users are never root, so this does not apply here. You could make this a (default) parameter, `user_is_root` and then handle that in the function, iff you want to make the function "universal".

Exactly, universal function. This is what I am telling the whole time. Once for all cases. If not, you are typing future pain to maintain code.


> You override the default assert_screen timeout, which is bad practice.

It is a draft, use the default.


> From the function name, I can not see what the function is supposed to do - `handle` is obviously wrong, as it does not "handle" anything. Also, 'generic-desktop' is valid here as well.

If my assumption about how the popup works, generic-desktop should not be part of the function.
It is a draft, you can choose a better name.


> When you got everything right, you end up with a complicated 10+ lines function which makes the tests actually harder to read, just to save 1 line each in 2 tests.

It is not to save space, it is to save future pain, because there are a lot of places where this needs to be handled.


> **If** there is a common pattern, repeated in several places, I am all for abstracting it away. Doing it upfront is IMHO not a good idea.

Sure, let's remake a root-console, bootloader, grub_test, first_boot, INSTALL_ONLY, etc. horror story ;)
They all started by just adding 2 lines of code, until each person from each team reapplied the same logic in different places so that we have mixed behavior everywhere.

Reply:
btw, I started to beg for a universal function since I am seeing at least 3 tickets per week created about this in different scenarios: https://progress.opensuse.org/issues/56267

Reply:
> btw, I started to beg for a universal function since I am seeing at least 3 tickets per week created about this in different scenarios: https://progress.opensuse.org/issues/56267

Although these are all related to the openSUSE Welcome dialog, the problems are different:

- Closing of the dialog not working reliably - this is an implementation detail, it took several attempts to get it reliable, but the general logic stayed the same. The opensuse_welcome test is used by all affected modules to handle this.

- opensuse_welcome not added in all places where required - doesn't matter if you have to add opensuse_welcome to the tests to run or add a function call, something you *have* to add.

Reply:
> Sure, let's remake a root-console, bootloader, grub_test, first_boot, INSTALL_ONLY, etc. horror story ;)
> They all started by just adding 2 lines of code, until each person from each team reapplied the same logic in different places so that we have mixed behavior everywhere.

Just because the same two lines are added does not mean everything is the same.

The tests are waiting for a specific state. Depending on the test, there are several possible valid states, each one denoted by a set of needles. Eventually running openSUSE Welcome extends the set of valid states.

**After** the state has been reached it is handled accordingly.

This PR **does not** add any handling of the openSUSE Welcome dialog, it just marks the needle as a valid outcome of previous actions. There is nothing special to **do** here, just logout as previously and be done with it.

Reply:
> > btw, I started to beg for a universal function since I am seeing at least 3 tickets per week created about this in different scenarios: https://progress.opensuse.org/issues/56267
> 
> Although these are all related to the openSUSE Welcome dialog, the problems are different:

I see the same popup and the same handling, closing it and checking that it is closed. Where is the difference?


>     * Closing of the dialog not working reliably - this is an implementation detail, it took several attempts to get it reliable, but the general logic stayed the same. The opensuse_welcome test is used by all affected modules to handle this.

I agree, for this reason I didn't wanted to consider the job as done and keep it open as https://progress.opensuse.org/issues/55661
As I mentioned in https://progress.opensuse.org/issues/56963#note-6

> 
>     * opensuse_welcome not added in all places where required - doesn't matter if you have to add opensuse_welcome to the tests to run or add a function call, something you _have_ to add.

How will you handle modules that switch users within the module? Or even modules that create users and then switch to it? You cannot invoke a module, but you can invoke a function.

Reply:
> Just because the same two lines are added does not mean everything is the same.

I cannot agree to just workaround the popup and pass the hot potato to the next one. The only way I can see this popup handled properly, is to close it the first time it appears for each logged in user and don't show on startup.


> The tests are waiting for a specific state. Depending on the test, there are several possible valid states, each one denoted by a set of needles. Eventually running openSUSE Welcome extends the set of valid states.
>
> This PR **does not** add any handling of the openSUSE Welcome dialog, it just marks the needle as a valid outcome of previous actions. There is nothing special to **do** here, just logout as previously and be done with it.

I agree with the goal, not with the method. The expected state should be done properly, and not as a workaround accepting any open popup and passing the hot potato to the next one.

Reply:
> > Although these are all related to the openSUSE Welcome dialog, the problems are different:
> 
> I see the same popup and the same handling, closing it and checking that it is closed. Where is the difference?

Because there is a difference between handling it incorrectly/unreliably and not handling it at all.

Reply:
> How will you handle modules that switch users within the module? Or even modules that create users and then switch to it? You cannot invoke a module, but you can invoke a function.

How will you **run** modules from within another modules? Iff you wan't to run a test, you have to finish/leave the "login as other user" module anyway.

Reply:
> > Just because the same two lines are added does not mean everything is the same.
> 
> I cannot agree to just workaround the popup and pass the hot potato to the next one. The only way I can see this popup handled properly, is to close it the first time it appears for each logged in user and don't show on startup.

There **is no workaround**. The appearing dialog is a valid result of logging in, as is "generic-desktop".

Eventually closing the dialog is another step.

> I agree with the goal, not with the method. The expected state should be done properly, and not as a workaround accepting any open popup and passing the hot potato to the next one.

The expected state is done properly. Pushing more and more stuff into a module, open-coded or via function calls makes the module harder to understand.

Take e.g. the unreliability of the opensuse_welcome module. If this were just function calls, you would get failures in at least 4 different modules, all because the same function call failed. But as it is a separate module, you can see already from the overview page which part is failing.